File: src/lock.rs
use std::fs;
use std::path::PathBuf;

const LOCK_DIR: &str = "/tmp/tncli";

fn lock_path(session: &str, service: &str) -> PathBuf {
    PathBuf::from(format!("{LOCK_DIR}/{session}_{service}.lock"))
}

pub fn ensure_dir() {
    let _ = fs::create_dir_all(LOCK_DIR);
}

pub fn acquire(session: &str, service: &str) {
    ensure_dir();
    let _ = fs::write(lock_path(session, service), format!("{}", std::process::id()));
}

pub fn release(session: &str, service: &str) {
    let _ = fs::remove_file(lock_path(session, service));
}

pub fn release_all(session: &str) {
    if let Ok(entries) = fs::read_dir(LOCK_DIR) {
        let prefix = format!("{session}_");
        for entry in entries.flatten() {
            let name = entry.file_name();
            let name = name.to_string_lossy();
            if name.starts_with(&prefix) && name.ends_with(".lock") {
                let _ = fs::remove_file(entry.path());
            }
        }
    }
}

File: internal/lock/lock.go
package lock

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

const lockDir = "/tmp/tncli"

func lockPath(session, service string) string {
	return filepath.Join(lockDir, fmt.Sprintf("%s_%s.lock", session, service))
}

func EnsureDir() {
	_ = os.MkdirAll(lockDir, 0o755)
}

func Acquire(session, service string) {
	EnsureDir()
	_ = os.WriteFile(lockPath(session, service), []byte(fmt.Sprintf("%d", os.Getpid())), 0o644)
}

func Release(session, service string) {
	_ = os.Remove(lockPath(session, service))
}

func ReleaseAll(session string) {
	entries, err := os.ReadDir(lockDir)
	if err != nil {
		return
	}
	prefix := session + "_"
	for _, e := range entries {
		name := e.Name()
		if strings.HasPrefix(name, prefix) && strings.HasSuffix(name, ".lock") {
			_ = os.Remove(filepath.Join(lockDir, name))
		}
	}
}
